### PR TITLE
JAMES-2586 ADR for Posgres mailbox tables structure

### PR DESCRIPTION
**Pull request by hungphan227**

### CODE REVIEW

File: src/adr/0071-postgresql-mailbox-tables-structure.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+
+Date: 2023-12-14
+
+## Status
+
+Implemented
+
+## Context
+
+Apache James was integrated with Postgresql database. The table structure for mailbox was created to store mailbox data in Postgresql database.
```
Comment (chibenwa) [inline src/adr/0071-postgresql-mailbox-tables-structure.md:11]: Why Past tense?

I propose: 

```
As described in [ADR-70](link), we are willing to provide a Postgres implementation for Apache James. 
The current document is willing to detail the inner working of the mailbox of the target implementation.
```

File: src/adr/0071-postgresql-mailbox-tables-structure.md [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+## Decision
+
+![diagram for mailbox tables](img/adr-71-mailbox-table-diagram.png)
```
Comment (chibenwa) [inline src/adr/0071-postgresql-mailbox-tables-structure.md:15]: The diagram is nice but it do not replace plain english with a table list.

We are missing annotations on it, annotations is part of the mailbox.

I would also like to have the word **normalized** explicitly mentionned: our table structure is mostly normalized which mitigates storage costs and ease achiving consistency. Indexes are used for key query in order to improve performance.

Also have a word about constraintd

File: src/adr/0071-postgresql-mailbox-tables-structure.md [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+Indexes in table message_mailbox: 
+- message_mailbox_message_id_index (message_id)
+- mailbox_id_mail_uid_index (mailbox_id, message_uid)
+- mailbox_id_is_seen_mail_uid_index (mailbox_id, is_seen, message_uid)
+- mailbox_id_is_recent_mail_uid_index (mailbox_id, is_recent, message_uid)
+- mailbox_id_is_delete_mail_uid_index (mailbox_id, is_deleted, message_uid)
+
+Indexes are used to find records faster
+
+For some fields, hstore data type are used. Hstore is key-value hashmap data structure. It is useful in some cases. For example, the database needs to store a list of usernames and corresponding rights for a mailbox. By applying hstore, acl that belongs to a specific mailbox and username could be retrieved quickly.
```
Comment (chibenwa) [inline src/adr/0071-postgresql-mailbox-tables-structure.md:26]: Why do we use hstore?

A: this postgres specific datatypes allows us to model complex data types without the need for complex joins.

File: src/adr/0071-postgresql-mailbox-tables-structure.md [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
```
Comment (chibenwa) [inline src/adr/0071-postgresql-mailbox-tables-structure.md:27]: I would like a word on consistency: Using specific postgres construct like returning or conflict management we are providing a consistent mailbox implementation without the need to rely on transaction thus with limited need to trade performance for consistency.

File: src/adr/0071-postgresql-mailbox-tables-structure.md (hunk 1 above)
Comment (hungphan227) [reply src/adr/0071-postgresql-mailbox-tables-structure.md]: fixed

File: src/adr/0071-postgresql-mailbox-tables-structure.md (hunk 2 above)
Comment (hungphan227) [reply src/adr/0071-postgresql-mailbox-tables-structure.md]: fixed

File: src/adr/0071-postgresql-mailbox-tables-structure.md (hunk 3 above)
Comment (hungphan227) [reply src/adr/0071-postgresql-mailbox-tables-structure.md]: fixed

File: src/adr/0071-postgresql-mailbox-tables-structure.md (hunk 4 above)
Comment (hungphan227) [reply src/adr/0071-postgresql-mailbox-tables-structure.md:41]: fixed

**Review by chibenwa — APPROVED** [review]

File: src/adr/0071-postgresql-mailbox-tables-structure.md [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+Indexes in table message_mailbox: 
+- message_mailbox_message_id_index (message_id)
+- mailbox_id_mail_uid_index (mailbox_id, message_uid)
+- mailbox_id_is_seen_mail_uid_index (mailbox_id, is_seen, message_uid)
+- mailbox_id_is_recent_mail_uid_index (mailbox_id, is_recent, message_uid)
+- mailbox_id_is_delete_mail_uid_index (mailbox_id, is_deleted, message_uid)
+
+Indexes are used to find records faster
+
+The table structure is mostly normalized which mitigates storage costs and achieve consistency easily
```
Comment (quantranhong1999) [inline src/adr/0071-postgresql-mailbox-tables-structure.md:34]:
```suggestion
The table structure is mostly normalized which mitigates storage costs and achieves consistency easily.
```

File: src/adr/0071-postgresql-mailbox-tables-structure.md [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+- subscription
+
+Indexes in table message_mailbox: 
+- message_mailbox_message_id_index (message_id)
+- mailbox_id_mail_uid_index (mailbox_id, message_uid)
+- mailbox_id_is_seen_mail_uid_index (mailbox_id, is_seen, message_uid)
+- mailbox_id_is_recent_mail_uid_index (mailbox_id, is_recent, message_uid)
+- mailbox_id_is_delete_mail_uid_index (mailbox_id, is_deleted, message_uid)
+
+Indexes are used to find records faster
```
Comment (quantranhong1999) [inline src/adr/0071-postgresql-mailbox-tables-structure.md:32]:
```suggestion
Indexes are used to find records faster.
```

**Review by quantranhong1999 — APPROVED** [review]

**Pull request merged**